### ISSUE TITLE
Reconsider usage of getBeansOfType in Jackson auto-configuration

### ISSUE THREAD

**Issue by philwebb** [spring-projects/spring-boot#14686]
During #14666 there was some discussion about the use of `getBeansOfType`. It seems that that method may swallow `BeanCreationExceptions`:

```java
	public <T> Map<String, T> getBeansOfType(@Nullable Class<T> type, boolean includeNonSingletons, boolean allowEagerInit)
			throws BeansException {

		String[] beanNames = getBeanNamesForType(type, includeNonSingletons, allowEagerInit);
		Map<String, T> result = new LinkedHashMap<>(beanNames.length);
		for (String beanName : beanNames) {
			try {
				Object beanInstance = getBean(beanName);
				result.put(beanName, (beanInstance instanceof NullBean ? null : (T) beanInstance));
			}
			catch (BeanCreationException ex) {
				Throwable rootCause = ex.getMostSpecificCause();
				if (rootCause instanceof BeanCurrentlyInCreationException) {
					BeanCreationException bce = (BeanCreationException) rootCause;
					String exBeanName = bce.getBeanName();
					if (exBeanName != null && isCurrentlyInCreation(exBeanName)) {
						if (logger.isDebugEnabled()) {
							logger.debug("Ignoring match to currently created bean '" + exBeanName + "': " +
									ex.getMessage());
						}
						onSuppressedException(ex);
						// Ignore: indicates a circular reference when autowiring constructors.
						// We want to find matches other than the currently created bean itself.
						continue;
					}
				}
				throw ex;
			}
		}
		return result;
	}
```

It might be better for us to `getBeanNamesForType` and then call `getBean` which will not swallow the error.

**Comment by philwebb** [comment]
@artembilan Can you provide instructions for how we test this against the sample that was provided in #14666

**Comment by artembilan** [comment]
Thanks, @philwebb 

Well, after your fix in the #14666 , I'm not sure now what we are chasing.
The problem before was that with such a `continue` in the mentioned `getBeansOfType()` we didn't get an `EventDrivenConsumer` bean for the `@Transformer` declared in that sample.

To test that I've just modified a `main()` method of the sample like this:
```
public static void main(String[] args) {
        ConfigurableApplicationContext applicationContext = SpringApplication.run(Application.class, args);
        Map<String, EventDrivenConsumer> beansOfType = applicationContext.getBeansOfType(EventDrivenConsumer.class);
        if (beansOfType.size() < 2) {
              throw new IllegalStateException();
        }
        applicationContext.close();
}
```

There should be two of them: one for the global `errorChannel` and another for the `channel1` in that `@Transformer` declaration.

Such a `@Transformer` is parsed in the `TransformerAnnotationPostProcessor` and there we have a code like:
```
protected Object resolveTargetBeanFromMethodWithBeanAnnotation(Method method) {
	String id = resolveTargetBeanName(method);
	return this.beanFactory.getBean(id);
}
```
where the `method` is the mentioned `@Bean` in the user configuration.
And it looks like this `getBean()` causes an eager initialization for the dependency of an `ObjectMapper` in the bean method.
That, in turn calls all the "magic" in the `JacksonAutoConfiguration`.

I won't mind that we may have some problem in Spring Integration as well, but right now it isn't clear where to go since we don't have an easy way to reproduce anymore...

**Comment by philwebb** [comment]
I have a feeling this might be a non-issue now. If that's not the case we can reopen.
